crates/iceberg/src/io.rs

```diff
@@ -0,0 +1,426 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! File io implementation.
+//!
+//! File io is built on top of [`opendal`](https://docs.rs/opendal/latest/opendal/index.html), which already provided an abstraction over all kinds of storage.
+//!
+//! # How to build `FileIO`
+//!
+//! We provided a `FileIOBuilder` to build `FileIO` from scratch. For example:
+//! ```rust
+//! use iceberg::io::{FileIOBuilder, S3_ARGS_BUCKET, S3_ARGS_REGION};
+//!
+//! let file_io = FileIOBuilder::new("s3")
+//!     .with_arg(S3_ARGS_BUCKET, "test_bucket")
+//!     .with_arg(S3_ARGS_REGION, "us-east-1")
+//!     .build()
+//!     .unwrap();
+//! ```
+//!
+//! We also provided a convenient method to build `FileIO` from url:
+//!
+//! ```rust
+//! use iceberg::io::{FileIO, S3_ARGS_REGION};
+//! let file_io = FileIO::build_from_url("s3a://test_bucket/warehouse")
+//!     .unwrap()
+//!     .with_arg(S3_ARGS_REGION, "us-east-1")
+//!     .build()
+//!     .unwrap();
+//! ```
+//!
+//! # How to use `FileIO`
+//!
+//! Currently `FileIO` provides simple methods for file operations:
+//!
+//! - `delete`: Delete file.
+//! - `is_exist`: Check if file exists.
+//! - `new_input`: Create input file for reading.
+//! - `new_output`: Create output file for writing.
+
+use std::{collections::HashMap, sync::Arc};
+
+use crate::{error::Result, Error, ErrorKind};
+use futures::{AsyncRead, AsyncSeek, AsyncWrite};
+use opendal::{Operator, Scheme};
+use url::Url;
+
+/// Following are arguments for s3 operator.
+/// s3 root
+pub const S3_ARGS_ROOT: &str = "root";
```

Comment:
Maybe it's better for opendal to support something like `s3://bucket/path/to/file?region=abc&access_key_id=xxxx`?

Reply:
Maybe it's better for `opendal` to expose these parameters and provide documentation? It would be great for opendal to provide such a thing, but I'm not sure if it's useful in iceberg. In iceberg the location are provided from table metadata but doesn't contains credentials in parameters.

Reply:
> In iceberg the location are provided from table metadata but doesn't contains credentials in parameters.

So the credentials are always fetched from users input (via iceberg-rust's API or ENV)?

Reply:
I think so. For example, let's imagine how to write data file for table in s3. We get table base location for table metadata, e.g. s3://bucket/demo/s1/t1. Then we create output file for a.parquet unber table base location. In this case, the credentials should be provided from user input or env.

Reply:
In PyIceberg we have three options:

- Environment variables

export PYICEBERG_CATALOG__DEFAULT__URI=thrift://localhost:9083

- Local configuration file `~/.pyiceberg.yaml`

```yaml
catalog:
  prod:
    uri: http://rest-catalog/ws/
    credential: t-1234:secret
```

- Pass it in directly:

```python
from pyiceberg.catalog import load_catalog

catalog = load_catalog(
    "prod",
    **{
        "uri": "http://127.0.0.1:8181",
        "s3.endpoint": "http://127.0.0.1:9000",
        "py-io-impl": "pyiceberg.io.pyarrow.PyArrowFileIO",
        "s3.access-key-id": "admin",
        "s3.secret-access-key": "password",
    }
)
```

Also, the S3 library will also check `AWS_ACCESS_KEY_ID` etc.

Another important aspect is that the REST catalog can also inject configuration (such as vended credentials, regions etc).

Reply:
The `FileIOBuilder` provides convenient methods for setting configurations for `FileIO`, e.g. you can override arguments with `with_arg` or `with_args` method. As with loading configuration from different places(env, env file, constructor args), I think it should be delegated to other mechanisms and resolved in following pr.

Reply:
That makes sense.